### PR TITLE
Feature/handle authentication dependencies

### PR DESCRIPTION
**Pull request by quanpython**
With APIs that don't require authentication, FastAPI doesn't provide the feature to specify them in whitelist format, so I created a separate APIRouter for these APIs.
If you have a better idea please let me know.

### CODE REVIEW

**Comment by ReiHashimoto** [comment]
@quanpython
Thank you for your investigation.

As you say, [fastapi's global dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/global-dependencies/) does not seem to have whitelist options.

However, I found implementation using same routers.
https://stackoverflow.com/questions/73141350/override-global-dependency-for-certain-endpoints-in-fastapi

Instead of creating routers on each files in /studio/app/*/routers, pre-define routers without-auth, with-auth, with-admin-auth may be one of the solution.

Please do not start implementation until we decide.
But, please let us know if there are any concerns with the above implementation at this time.

**Comment by quanpython** [comment]
@ReiHashimoto
> However, I found implementation using same routers.

I see it is using spliting endpoints in two routers. Is there any mistake here?

> Instead of creating routers on each files in /studio/app/*/routers, pre-define routers without-auth, with-auth, with-admin-auth may be one of the solution.

Do you mean that using these pre-define routers for all endpoints? If so, the other parameters like `prefix` and `tag` will  need to be redefined.

**Comment by ReiHashimoto** [comment]
@itutu-tienday 
私の方でも検証してみましたが、結局のところ、以下のどちらかのパターンで進めることになりそうです。
要は事前に定義するのを何にするか、の問題ですね。
個人的には2のような実装がベターかとは思います。

1. 認証の有無などによってルーターを分ける
    ```python
    default_router=APIRouter()
    auth_router=APIRouter(dependencies=[Depends(get_admin_user)])
    admin_router=APIRouter(dependencies=[Depends(get_admin_user)])
    ```
    - 現在は各ファイルでrouterを定義しており、tagやprefixなどを一括で割り当てていますが、`auth_router.tag="some_tag"`のように各ファイルでオーバーライドができません。パスごとにタグやprefixを指定する必要があります。
    - さらに、結局事前定義したルータをimportして使用しているかをチェックする必要があり、ニーズに合わないと思います。
2. include_routerでdependenciesを指定するようにする
   ```python
   app.include_router(xxx.router, dependencies=[Depends(...)])
   ```
   app.include_router()でdependenciesを指定できます。
   __main_unit__.pyでdependenciesを指定するようにすれば、認証の設定忘れなどが一括で確認できそうです。

File: studio/__main_unit__.py [hunk 1]
```diff
@@ -44,6 +44,7 @@
 app.include_router(hdf5.router)
 app.include_router(nwb.router)
 app.include_router(roi.router)
+app.include_router(expdb.public_router)
```
Comment (itutu-tienday) [inline studio/__main_unit__.py:47]: @quanpython 
As stated in #117 requirements, we expect to apply authentication to all APIs (except public database, auth).
→ Currently, authentication does not seem to be applied across the board (experiments, run, etc.), so please apply it.

File: studio/app/common/routers/users_me.py [hunk 2]
```diff
@@ -6,7 +6,9 @@
 from studio.app.common.db.database import get_db
 from studio.app.common.schemas.users import SelfUserUpdate, User, UserPasswordUpdate
 
-router = APIRouter(prefix="/users/me", tags=["users/me"])
+router = APIRouter(
+    prefix="/users/me", tags=["users/me"], dependencies=[Depends(get_current_user)]
```
Comment (itutu-tienday) [inline studio/app/common/routers/users_me.py:10]: @quanpython 
Here `Depends(get_current_user)` is specified, but it seems that each API method also specifies the above Depends one by one.
-> #117 requirement, the purpose is to save labor for the above individual settings, but could the description format be simplified a bit more?

**Review by itutu-tienday — CHANGES_REQUESTED** [review]

File: studio/app/common/routers/users_me.py (hunk 2 above)
Comment (quanpython) [reply studio/app/common/routers/users_me.py]: @itutu-tienday For APIs that need to determine the information of the currently logged user or admin, a definition for each such method is required.

File: studio/__main_unit__.py (hunk 1 above)
Comment (quanpython) [reply studio/__main_unit__.py:51]: @itutu-tienday  I applied it

File: studio/app/common/routers/users_me.py (hunk 2 above)
Comment (itutu-tienday) [reply studio/app/common/routers/users_me.py]: @quanpython 
I'd like to confirm a few points,

- So you are saying that it seems difficult to simplify than this configuration in the current situation?
- Wouldn't setting [Depends(get_current_user)] for each include_router and each API cause the process to be executed twice? 
  - When I checked the operation, it does not seem to be executed twice, but is this the expected behavior?

Comment (quanpython) [reply studio/app/common/routers/users_me.py]: @itutu-tienday
> So you are saying that it seems difficult to simplify than this configuration in the current situation?

I think it is simple and reasonable enough. One dependency indicated in the API method is for retrieving the returned value, while the other, indicated in `include_router`, offers a summary for the router.
 
> Wouldn't setting [Depends(get_current_user)] for each include_router and each API cause the process to be executed twice?

If a dependency is declared multiple times for the same path operation, FastAPI will call it only once per request and cache the returned value.
Docs: https://fastapi.tiangolo.com/tutorial/dependencies/sub-dependencies/#using-the-same-dependency-multiple-times

Comment (itutu-tienday) [reply studio/app/common/routers/users_me.py]: @quanpython 
I understand your assumption and the FastAPI dependency specification.
We will merge with this implementation.

**Review by itutu-tienday — APPROVED** [review]

**Pull request merged**